design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+ which pulls out the struct fields.
+
+## Value/Return
+
+This functionality will make it easier for users to make end-end streaming apps using KSQL
+alone without having to have custom stages using Kafka Streams.
+
+We believe its a common case for users to have array or map like data structures as values
+in messages that they wish to explode. This new feature would allow them to do that in a pure KSQL
+app thus allowing them to create richer KSQL applications.
+
+## Public APIs
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
+UDAFs.
+
+## Design
+
+### General plumbing
+
+We will need the following general plumbing to support table functions:
+
+* In a similar way to how UDAFs are currently supported we will need to create a 
+`KsqlTableFunction` which is a wrapper around the actual table function (whether built-in or
+generated). We'll also need to augment the function registry to manage the addition and retrieval
+of table functions.
+* Augment the query analysis code to extract information relating to any table functions
+in the query. We can do this either by extending the existing query analyzer or perhaps creating
+a new `TableFunctionAnalyzer` in a similar way to how aggregate analysis uses its own 
+`AggregateAnalyzer`.
+* Augment the building of logical plan by building a new node representing a FlatMap stage of the
+stream processing in the case there are table functions in the SELECT.
+* Augment the execution plan logic by adding new nodes in the execution plan corresponding to any
+extra FlatMap stage required by the query.
+* Amend the columns used in the final projection to reflect any generated columns created by a 
+FlatMap stage.
+* Implement the actual FlatMap logic. For each row, any table functions on that row will be evaluated.
+The number of rows returned from the FlatMap will be equal to the greatest of the number of rows
+returned from any of the table functions for that row. Any non table function values for the row will
+be duplicated for each row. For any table function that returned fewer than the greatest number of rows
+the values for that columnn will be `null`. For performance reasons, the FlatMap logic should cache any
+ table functions for the columns to avoid them being looked up each time.
+
+### Implement set of built in table functions
+
+We will implement the following built in table functions:
+
+* `EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a column
+of type `V` holding the value of the element
+* `EXPLODE(MAP<K, V>)` -> Results in `N` rows (where `N` is number of elements in map) each with a column
+holding a `STRUCT<key K, value V>` that contains the key and value.
+* `INDEX_EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a
```

Comment:
This is a cool idea, although I'm wondering if we'd be pulling too much functionality into `explode` with this (I understand it's a separate function, but nevertheless). That is, could there be a way to achieve this result using an intuitive composition of simpler functions instead? For example, we could have a `series` (table!) function that could be used to enumerate an array like so:

```sql
-- Example event:
-- {"arr": [0, 1, 2]}

SELECT series(0, len(arr)) AS index, explode(arr) FROM stream EMIT CHANGES;
index | explode
----------------
  0   |   0
  1   |   1
  2   |   2
```
Or even something that wraps the above array length check:
```sql
SELECT enumerate(arr), explode(arr) ...
```
In general I believe we'll be able to provide users with greater flexibility by solving problems like these using compositions of simple functions. Good composability is also conducive to solving unforeseen problems/use cases that we didn't necessarily design for.

Reply:
This particular function is borrowed from Hive (pos_explode). I agree with the point about composability but we already have an explode method which doesn't provided indexes. And I think functions like series() would be a good idea, so they could just be used together to achieve the same ends.
I think sometimes though it's useful to provide some more denormalised functions for commonly used, use cases. I'm guessing that getting the index is going to be one of those, but it's just a guess! ;) It's the same kind of reasoning of why HashMap in Java has an entries() collection as well keys() and values() ones. It's not strictly necessary but it's so often the case you want to iterate the key and value together that it can lead to clearer code. Having said that I'm not hung up on this function and will remove it if people don't think it's useful.

Reply:
IMHO I see the benefit of pos_explode, but if given the choice between `series` _or_ `pos_explode` I'd take the former any day.  

So maybe we should consider either replacing `pos_explode` in this doc with `series` or adding `series` as well, but not having `pos_explode` without `series`.